### PR TITLE
feat(imgproc): add linear-time euclidean distance transform

### PR DESCRIPTION
**Pull request by zeel991**
## 📝 Description

**Fixes:** https://github.com/kornia/kornia-rs/issues/275

## 🛠️ Changes Made
- [x] Implemented O(N) linear-time Euclidean Distance Transform using the Felzenszwalb & Huttenlocher algorithm.
- [x] Updated function signatures to use `ImageAllocator` and `CpuAllocator` to ensure compatibility with the latest library changes.
- [x] Fixed the legacy `distance_transform_vanilla` implementation by replacing the `ndarray` dependency with standard `Vec` and fixing an off-by-one loop error.
- [x] Added documentation for public functions to satisfy linting requirements.

---

## 🧪 How Was This Tested?
- [x] **Unit Tests:** Added `test_accuracy_vs_vanilla` to verify the O(N) implementation matches the O(N^2) ground truth. Added `distance_transform_smoke` for basic sanity checks.
- [x] **Manual Verification:** Ran `cargo test -p kornia-imgproc` locally to ensure all tests pass and compilation succeeds without warnings.
- [x] **Performance/Edge Cases:** Validated that the new algorithm handles image boundaries correctly (fixed previous index errors) and performs in linear time O(N) compared to the previous O(N^2) approach.

---

## 🕵️ AI Usage Disclosure
*Check one of the following:*
- [ ] 🟢 **No AI used.**
- [x] 🟡 **AI-assisted:** I used AI for boilerplate/refactoring but have manually reviewed and tested every line.
- [ ] 🔴 **AI-generated:** (Note: These PRs may be subject to stricter scrutiny or immediate closure if the logic is not explained).

---

## 🚦 Checklist
- [x] I am assigned to the linked issue (required before PR submission)
- [x] The linked issue has been approved by a maintainer
- [x] This PR strictly implements what the linked issue describes (no scope creep)
- [x] I have performed a **self-review** of my code (no "ghost" variables or hallucinations).
- [x] My code follows the existing style guidelines of this project.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [x] I have added tests that prove my fix is effective or that my feature works.

---

## 💭 Additional Context(Test Output)
Screenshot from 2026-01-22 12-46-53

### CODE REVIEW

**Comment by sidd-27** [comment]
Please Provide benchmarking as well as comparison against other libraries eg open CVs implementation

**Comment by zeel991** [comment]
> Please Provide benchmarking as well as comparison against other libraries eg open CVs implementation

##  Performance Benchmarks

### 1. Internal Comparison 
Comparison between the old O(N²) brute-force implementation and the new O(N) implementation. The speedup is massive and grows with image size.

| Image Size | Implementation | Time (avg) | Speedup |
| :--- | :--- | :--- | :--- |
| **64x64** | Vanilla (Old) | 9.23 ms | - |
| **64x64** | **New O(N)** | **0.056 ms** | **~165x Faster** |
| **128x128** | Vanilla (Old) | 130.35 ms | - |
| **128x128** | **New O(N)** | **0.218 ms** | **~598x Faster** |
| **512x512**| **New O(N)** | **3.83 ms** | (Old timed out) |
| **1024x1024**| **New O(N)** | **19.45 ms** | (Old timed out) |

### 2. External Comparison 
Comparison against `cv::distanceTransform` using the `opencv` crate.

| Image Size | Library | Time (avg) | Notes |
| :--- | :--- | :--- | :--- |
| **512x512** | OpenCV (C++) | 0.385 ms | SIMD optimized |
| **512x512** | **kornia-rs** | **3.83 ms** | Pure Rust (Scalar) |
| **1024x1024** | OpenCV (C++) | 1.52 ms | SIMD optimized |
| **1024x1024** | **kornia-rs** | **19.45 ms** | Pure Rust (Scalar) |

*Note: Benchmarks run on Ryzen 5 / Dell G15.*

**Comment by sidd-27** [comment]
Nice! Is there a way to get closer to the opencv speed? try to use the Wide crate if possible for SIMD

**Comment by zeel991** [comment]
> Nice! Is there a way to get closer to the opencv speed? try to use the Wide crate if possible for SIMD

I considered using the wide crate for SIMD, but I'm concerned about the loop in the parabolic intersection step. Since the number of backtracks is data-dependent, I feel direct vectorization will suffer from heavy branch divergence

**Comment by sidd-27** [comment]
then how does openCV do it?

**Comment by zeel991** [comment]
> then how does openCV do it?

OpenCV achieves its speed primarily through multithreading (TBB/OpenMP) combined with handwritten SIMD intrinsics.

To bridge the gap without the complexity of manual vectorization (and avoiding branch divergence issues with wide), I implemented a Rayon (Multithreading) + Transpose approach. This matches OpenCV's parallel execution model and optimizes memory access patterns.

The performance improvement is significant:

Implementation | Time (1024x1024) | Improvement | vs OpenCV
-- | -- | -- | --
Rust (Single-Threaded) | 19.50 ms | Baseline | ~13.1x Slower
Rust (Rayon + Transpose) | 4.58 ms | 76.5% Faster (4.2x) | ~3.1x Slower
OpenCV  | 1.49 ms | - | Baseline

I can commit the multithreaded algorithm as well , if the approach is acceptable

**Comment by sidd-27** [comment]
Sure go ahead with the multi threaded version

**Comment by zeel991** [comment]
@sidd-27 do let me know if there are any more changes required

**Comment by sidd-27** [comment]
LGTM, I'll let others take a look before we can merge

**Comment by zeel991** [comment]
hello @edgarriba, can you please review the PR, and let me know if any more changes are required

File: crates/kornia-imgproc/src/distance_transform.rs [hunk 1]
```diff
@@ -1,107 +1,291 @@
...
+        .for_each(|(g, &s)| {
+            if s > 0.0 {
+                *g = 0.0;
+            }
+        });
+
+    // Transform along Rows (Parallel)
+    // We allocate thread-local scratch buffers to avoid repeated allocations.
+    grid.par_chunks_mut(width).for_each_init(
+        || (vec![0.0; width], vec![0usize; width], vec![0.0; width + 1]),
```
Comment (edgarriba) [inline crates/kornia-imgproc/src/distance_transform.rs]: probably we could make this function as a struct that has this vec as cache to avoid runtime allocation, that will give some improvement

File: crates/kornia-imgproc/src/distance_transform.rs [hunk 2]
```diff
@@ -1,107 +1,291 @@
...
+    let mut k = 0;
+    v[0] = 0;
+    z[0] = -INF;
+    z[1] = INF;
+
+    for q in 1..n {
+        loop {
+            let r = v[k];
+            // Calculate intersection of parabola 'r' and 'q'
+            debug_assert!(q != r, "q and r should never be equal");
```
Comment (edgarriba) [inline crates/kornia-imgproc/src/distance_transform.rs]: @sidd-27 do we want this ? let's get into claude/agents md this patterns

Comment (sidd-27) [reply crates/kornia-imgproc/src/distance_transform.rs]: from what i know, asserts do help the compiler, ill add to the md file

File: crates/kornia-imgproc/src/distance_transform.rs [hunk 3]
```diff
@@ -1,107 +1,291 @@
...
-    ((x1[0] - x2[0]).powi(2) + (x1[1] - x2[1]).powi(2)).sqrt()
+// We use a large finite number instead of f32::INFINITY to avoid NaN generation
+// during arithmetic operations (INF - INF = NaN).
+// 1e20 is sufficient for images up to ~1e10 pixels in dimension.
+const INF: f32 = 1e20;
+
+/// Computes the Euclidean Distance Transform of a binary image.
+///
+/// This implementation is highly optimized:
+/// 1. Uses the O(N) linear-time algorithm by Felzenszwalb & Huttenlocher (Distance Transforms of Sampled Functions, 2012).
```
Comment (sidd-27) [inline crates/kornia-imgproc/src/distance_transform.rs]: also remove these comments, in place of something simpler

**Comment by zeel991** [comment]
> to many runtime allocated vectors

I've optimized this to use just two pre-allocated buffers and transpose between them to minimize allocations

**Review by sidd-27 — COMMENTED** [review]
simplify comments, keep only useful ones

File: crates/kornia-imgproc/src/distance_transform.rs (hunk 3 above)
Comment (zeel991) [reply crates/kornia-imgproc/src/distance_transform.rs]: done!

**Comment by zeel991** [comment]
hey @sidd-27 ,

I have implemented a GPU-accelerated Distance Transform using the Felzenszwalb-Huttenlocher algorithm via CubeCL. The implementation uses a separable 1D approach: Column Pass → Transpose → Row Pass. This transpose strategy ensures optimal coalesced memory access for both passes.

Resolution| Vanilla CPU | Current Approach CPU | GPU Cold | GPU Hot | OpenCV |
|---------|-------------|----------------------|----------|---------|--------|
|64 x 64 | 14.696 ms   | 45.505 µs | 365.67 µs | 323.79 µs | 9.3208 µs | 
|128 x 128 | 160.20 ms | 233.62 µs | 570.01 µs | 477.32 µs | 28.646 µs | 
|512 x 512 | - | 4.6815 ms | 4.1913 ms | 2.0031 ms | 1.2314 ms|
|1024 x 1024 | - | 29.942 ms | 23.779 ms | 4.0951 ms | 4.0874 ms |
|2048 x 2048 | - | 203.17 ms | 293.33 ms | 8.9625 ms | 20.894 ms |

The difference between GPU Cold and GPU Hot benchmarks is a matter of measurement scope:

- GPU Cold includes the "setup tax": initializing the DistanceTransformer, allocating VRAM, and the initial JIT-compilation of kernels.
- GPU Hot reflects raw execution throughput: it reuses pre-allocated buffers and cached kernels, which is the realistic performance for persistent pipelines.

At $2048 \times 2048$, the GPU Hot implementation is ~2.3x faster than OpenCV and ~22x faster than the current CPU version. The Cold start remains high at large resolutions because the overhead of memory orchestration and hardware "wake-up" significantly outweighs the actual computation time.

**Comment by sidd-27** [comment]
LGTM, lets hold the gpu work for a bit, a decent amount of benchmarking and experimentation to be done there, otherwise looks good

**Comment by sidd-27** [comment]
fix lint errors

File: crates/kornia-imgproc/src/distance_transform.rs [hunk 4]
```diff
@@ -24,84 +30,180 @@ pub fn distance_transform_vanilla(image: &Image<f32, 1>) -> Image<f32, 1> {
...
+            f_col[y] = grid[y * width + x];
+        }
+        distance_transform_1d(&f_col, &mut d_col, &mut v_col, &mut z_col);
+        for y in 0..height {
+            grid[y * width + x] = d_col[y];
+        }
+    }
+
+    // Transform along Rows
+    let mut f_row = vec![0.0; width];
```
Comment (edgarriba) [inline crates/kornia-imgproc/src/distance_transform.rs]: probably the function should be hosted into a struct to hold this buffers, which can be allocated only once or if input sizes change. this will speedup things

Comment (zeel991) [reply crates/kornia-imgproc/src/distance_transform.rs]: can you review once again please

**Review by edgarriba — COMMENTED** [review]
the felzenszwalb distance transform is a good algorithm choice — O(n) per dimension.

review:

1. **good**: `DistanceTransformExecutor` with reusable buffers avoids repeated allocations. this is the pattern we want.

2. **good**: parallel row processing for the first pass, then transpose + parallel columns. smart.

3. **benchmarks**: you have opencv comparison which is great. can you share the actual numbers? specifically for 1024x1024 and 2048x2048 — how does kornia compare to opencv's `DIST_L2`?

4. **the `distance_transform_vanilla` function**: it's O(n^4) which is fine as a test reference but make sure it's not exported as public API (or at least clearly marked as test-only). i see it's `pub` — should be `pub(crate)` or behind a test feature.

5. the `transpose_map` function — is this doing a full matrix transpose? for large images that's a lot of cache misses. have you considered cache-oblivious blocking?

6. **`bench_draw` was replaced** — the bench entry was renamed from `bench_draw` to `bench_distance_transform`. was `bench_draw` removed intentionally? make sure we didn't lose an existing benchmark.

overall solid implementation. address the public API concern and share benchmark numbers.

**Comment by zeel991** [comment]
> the felzenszwalb distance transform is a good algorithm choice — O(n) per dimension.
> 
> review:
> 
> 1. **good**: `DistanceTransformExecutor` with reusable buffers avoids repeated allocations. this is the pattern we want.
> 2. **good**: parallel row processing for the first pass, then transpose + parallel columns. smart.
> 3. **benchmarks**: you have opencv comparison which is great. can you share the actual numbers? specifically for 1024x1024 and 2048x2048 — how does kornia compare to opencv's `DIST_L2`?
> 4. **the `distance_transform_vanilla` function**: it's O(n^4) which is fine as a test reference but make sure it's not exported as public API (or at least clearly marked as test-only). i see it's `pub` — should be `pub(crate)` or behind a test feature.
> 5. the `transpose_map` function — is this doing a full matrix transpose? for large images that's a lot of cache misses. have you considered cache-oblivious blocking?
> 6. **`bench_draw` was replaced** — the bench entry was renamed from `bench_draw` to `bench_distance_transform`. was `bench_draw` removed intentionally? make sure we didn't lose an existing benchmark.
> 
> overall solid implementation. address the public API concern and share benchmark numbers.


**3. Benchmark numbers (1024x1024 and 2048x2048):**

| Size       | kornia_cpu | opencv (DIST_MASK_5) | opencv (DIST_MASK_PRECISE) | 
|------------|------------|----------------------|----------------------------|
| 1024x1024  | 4.63 ms    | 1.50 ms              | 1.57 ms                    | 
| 2048x2048  | 25.06 ms   | 6.61 ms              | 7.11 ms                    | 

The main bottleneck is likely cache pressure from strided writes in `transpose_map`. 

**4. `distance_transform_vanilla` visibility**
I will change to `pub(crate)`.

**5. `transpose_map` cache misses**
Yes, the current implementation does naive strided writes which are cache-unfriendly 
at large sizes. This is likely the primary driver of the growing gap vs OpenCV.

**6. `bench_draw` → `bench_distance_transform` rename**
I mistakenly removed `bench_draw` when adding `bench_distance_transform`. I'll restore `bench_draw` in the next commit.

**Comment by edgarriba** [comment]
@zeel991 thanks for the update! The `DistanceTransformExecutor` struct with reusable buffers is exactly the pattern we want — good refactor.

A few things to check off before merge:
1. **Benchmark numbers** — you mentioned the executor is ready; can you share actual benchmark output comparing kornia vs OpenCV for 1024×1024 and 2048×2048? This is important to have in the PR for reference.
2. **GPU path** — as @sidd-27 noted, let us hold that for a follow-up PR to keep scope contained.
3. **CI** — make sure everything is green.

If benchmarks look good, LGTM — @sidd-27 / @cjpurackal can do the final review and merge.

**Comment by zeel991** [comment]
> @zeel991 thanks for the update! The `DistanceTransformExecutor` struct with reusable buffers is exactly the pattern we want — good refactor.
> 
> A few things to check off before merge:
> 
> 1. **Benchmark numbers** — you mentioned the executor is ready; can you share actual benchmark output comparing kornia vs OpenCV for 1024×1024 and 2048×2048? This is important to have in the PR for reference.
> 2. **GPU path** — as @sidd-27 noted, let us hold that for a follow-up PR to keep scope contained.
> 3. **CI** — make sure everything is green.
> 
> If benchmarks look good, LGTM — @sidd-27 / @cjpurackal can do the final review and merge.

| Size       | vanilla      | kornia_cpu  | opencv | ratio (kornia vs opencv) |
|------------|--------------|-------------|----------------------------|--------------------------|
| 64x64      | 10.106 ms    | 91.305 µs   | 23.922 µs                  | ~3.8x            |
| 512x512    | -            | 1.1533 ms   | 391.72 µs                  | ~2.9x            |
| 1024x1024  | -            | 4.6337 ms   | 1.5732 ms                  | ~2.9x            |
| 2048x2048  | -            | 25.057 ms   | 7.1123 ms                  | ~3.5x            |

I'll commit the one with corrected TOML file for restoring the draw benchmarks

**Review by sidd-27 — APPROVED** [review]
